lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -132,79 +132,134 @@ sub schedule {
-            next;
+
+        # check whether these jobs can actually be scheduled
+        my $last_worker_class;
+        my $skip = 0;
+        for my $job (@jobs) {
+            my $job_id = $job->id;
+            if ($job->state ne SCHEDULED) {
+                log_debug("Job $job_id no longer scheduled, skipping");
+                $skip = 1;
+                last;
+            }
+            my $worker_class = $job->settings->find({key => 'WORKER_CLASS'});
+            next unless $worker_class;
+            if ($worker_class ne ($last_worker_class //= $worker_class)) {
+                log_debug("Job $job_id needs same worker class as directly chained dependency"
+                      . " (has $worker_class, other chained dependencies have $last_worker_class), skipping");
+                $skip = 1;
+                last;
+            }
         }
+        next if $skip;
+
         my $res;
         try {
-            $res = $job->ws_send($worker);    # send the job to the worker
+            if ($actual_job_count > 1) {
+                $res
+                  = $self->_assign_multiple_jobs_to_worker(\@jobs, $worker, $directly_chained_job_hierarchy, $job_ids);
+            }
+            else {
+                $res = $jobs[0]->ws_send($worker);
+            }
             die "Failed contacting websocket server over HTTP" unless ref($res) eq "HASH" && exists $res->{state};
         }
         catch {
-            log_debug("Failed to send data to websocket :( bummer! Reason: $_");
+            log_debug("Failed to send data to websocket, reason: $_");
         };
 
-        # We succeded dispatching the message
         if (ref($res) eq "HASH" && $res->{state}->{msg_sent} == 1) {
-            log_debug("Sent job '" . $allocated->{job} . "' to worker '" . $allocated->{worker} . "'");
-            my $scheduled_state;
+            log_debug("Sent job(s) '$job_ids_str' to worker '$worker_id'");
+
+            # associate the worker to the job, so the worker can send updates
             try {
-                # We associate now the worker to the job, so the worker can send updates.
-                if ($job->set_assigned_worker($worker)) {
-                    push(@successfully_allocated, {job => $allocated->{job}, worker => $allocated->{worker}});
+                if ($actual_job_count > 1) {
+                    my %worker_assignment = (
+                        state              => ASSIGNED,
+                        t_started          => undef,
+                        assigned_worker_id => $worker_id,
+                    );
+                    $_->update(\%worker_assignment) for @jobs;
+                    $worker->update({job_id => undef});
+                    # note: The job_id on the worker's table makes no sense for multiple job assignments so leave
+                    #       it just blank.
+                    # FIXME: Adjust all places where this relation is used.
```

Comment:
I'd remove the relation first - while removing it you might realize what we missed.

Reply:
Ok, I'll try to remove the relation completely.

Reply:
I'm now using that relation to keep track of the currently executed job of the worker.